### DISCUSSION TITLE
WT32-eth01 with Tasmota and Neopool - firmware working - but which pinout and configuration for modbus to use ?

### DISCUSSION THREAD

**Discussion by ollyins** [arendst/Tasmota#18691] [Q&A]
Hello, i am searching around to find a solution for my problem, to connect the ttl to modbus converter to my wt32, like i did before with a normal ESP32 (using GPIO1 TX and GPIO3 for RX). On the WT32 i have the RX0 and TX0 pins which i used for flashing and i also have a named RXD port which is described as  io5 and TXD which should be io17. But my problem is, which ports i connect in the Tasmota configuration to Neopool RX and Neopool TX because i have no gpio17 port in the wt32 template which is also here mentioned in other threats! I also tried to use RX0 and TX0 which are the ports for flashing the device... but its also not clear which pin numbers/ports this are in the template - i tried gpio1 and 3.. buzt that seems to be wrong.
here the pinout from the wt32 and also the tasmota configuratio posibility, So it would be nice to get some information exactly regarding this pinout and config posibility from the tasmota layout! 
Which pins would be the RX0 and TX0 used for flashing? are they possible to configure in the template or not?

![grafik](https://github.com/arendst/Tasmota/assets/108992630/5e0e7940-ec59-4937-b70e-e55098888cbb)

**Comment by barbudor** [comment]
The board pinout seems to be pretty clear
The pins in grey at the top are the one used for flashing and debug / serial console
Although it shows a little error as TXD is GPIO1, not GPIO0

I suggest you leave those for that purpose and connect NeoPool to any other couple of GPIO available.
On ESP32 you are free to select any GPIO for any of the 3 available serial ports (UART)

GPIO17 is already defined in the template as Ledlink so you can either pickup another GPIO for Neopool, or re-assign that GPIO to User in the template so you can use it for NeoPool. That means that the on-board led will flash according to NeoPool serial activity

**Comment by shikousa** [comment]
@ollyins What is your final configuration for your TTL to RS485/Modbus converter?
In Tasmota V13 the GPIO17 is also not available. Can I use any other PIN like GPIO12 and GPIO14 for connecting the TTL to RS485 converter?

The WT32-ETH01 is supplied by a 5V power supply over the "5V power supply" PIN. Is it then possible to use the "3.3V power supply" PIN to supply the TTL to RS485 converter?

**Reply by curzon01** [reply]
> The WT32-ETH01 is supplied by a 5V power supply over the "5V power supply" PIN.

doesn't matter, 5V is stepped down to 3,3 V onboard

> Is it then possible to use the "3.3V power supply" PIN to supply the TTL to RS485 converter?

should be, but your RS485 converter must also work with 3.3V (if it has a MAX485, it won't work)

**Comment by shikousa** [comment]
I have changed LedLink_i from GPIO17 to GPIO0, but it is still not available.
Configure Template:
![grafik](https://github.com/arendst/Tasmota/assets/3395690/e3bfe4d3-5be7-4d66-96dd-70e3dae809a6)

Configure Module:
![grafik](https://github.com/arendst/Tasmota/assets/3395690/f130255b-e668-4595-9e5c-c00c0dc34843)

Is GPIO0 ok and what do I need to change in addition, so that I can select GPIO17 in Configure Module.

**Reply by sfromis** [reply]
What are you trying to do? Gpio 17 is connected to a LED on the board, and makes excellent sense as `LedLink_i`. If you don't want LedLink, you can unconfigure it (not essential to Tasmota operation), but any other usage of gpio 17 may conflict with the led. Other than the led on the pin, gpio 17 has no special status, and is not preferable to other pins. Still, one speciality with gpio 16+17 is that standard Tasmota configs will probe these for PSRAM available during boot, which may make it preferable to stay away from these pins.

With the way ESP32 works, there is no particular reason to want to use gpio 17 pin, but you have to select free pins to to avoid conflicts. And there are a lot of gpios reserved for Ethernet without this being directly reflected in the template.

**Comment by shikousa** [comment]
I would like to connect a TTL-RS485 converter und would like to use hardware UART. Therefore I need GPIO05 (RXD) and GPIO17 (TXD). What PINs would you recommend to use a TTL-RS485 converter?

**Reply by barbudor** [reply]
On ESP32 there is no dedicated pins for Uart
Any pin works and will be rerouted inside the chip

**Reply by sfromis** [reply]
The main concern is whether the pin is already used for something else, which is not straightforward to see with the Ethernet driver.

While I've not really worked what will or not work, a comment block in the driver does mentioned a list of dedicated/fixed pins.
https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_82_esp32_ethernet.ino#L23-L58
Not sure how "hard" that is, as gpio 17 is also on the list, but I'd be careful about trying to use those pins for other purposes, especially the RMII pins.

Also, on the original ESP32, gpios 34,35,36,39 are input-only.

It makes no difference what pin labels may happen to be used on the board, those are likely just to match the default firmware. Of course, when it is "just" pin numbers, they generally do make a lot of sense, but outside of that, it starts to be more arbitrary.

**Reply by shikousa** [reply]
What PINs would you recommend for UART when GPIO17 is not available? I only had seen in this project that they were using GPIO05 and GPIO17.

https://github.com/zigpy/zigpy/discussions/584

**Reply by barbudor** [reply]
Any free ones
AFAIK, the UART can be re-routed internally to any GPIOs

**Reply by shikousa** [reply]
The problem with the WT32-ETH01 is, that almost all available GPIOs are strapping PINs. I only see GPIO14 available. What about GPIO4, is this a strapping PIN or not. Reading the manual it seems to be the 6th one. Can GPIO4 be used like a regular GPIO?

Maybe the input-only PINs can also be used for RX.

**Reply by sfromis** [reply]
Not sure why you talk about lots of strapping pins, that's not the case. And even the few strapping pins **can** be used, as long as you make sure that the boot/power-up state is suitable.

From what I can find, these are the pins of WT32-ETH01, with gpio numbers instead of only opaque labels: 
![image](https://github.com/arendst/Tasmota/assets/47082390/b8522bc9-29c4-453f-972d-3ffd79765f36)

```
    1:   IO1 TX0     EN
    2:   IO3 RX0     GND
    3:   IO0         3V3
    4:   GND         EN
    5:   IO39        IO32 CFG
    6:   IO36        IO33 485_EN
    7:   IO15        IO5  RXD
    8:   IO14        IO17 TXD
    9:   IO12        GND
    10:  IO35        3V3
    11:  IO4         GND
    12:  IO2         5V
    13:  GND         LINK
```

I'm not saying that everything will work just so with the pins in the image, just that there are quite a few to try with, without seeing nearly everything as off-limits. Obviously, do test before committing to using some pin for some specific purpose.

Also, I observe that the templates page mentions that with Ethernet in use, I2C only works on gpio 32/33 as SCL/SDA.
https://templates.blakadder.com/wireless_tag_WT32-ETH01

**Comment by shikousa** [comment]
I have done some testing with the WT32-ETH01.  GPIO4 and GPIO14 are working. GPIO2 did not work, here I had a stuck in boot. Are there other PINs except GPIO4 and GPIO14 what could be used for UART (RX/TX) as I actually need 2 UARTs. Is it possible to use the Input only PINs 35,36,39 as RX? Then I would use GPIO4 and GPIO14 as TX.

**Reply by sfromis** [reply]
UART is not tied to any pins, any two are workable, as long as you are careful about how you use the few strapping pins, like state of gpio 0 and 2 being used to determine the boot mode. And of course not use the input-only pins for output.

There are quite a few pins to pick from:
```
    1:   IO1 TX0     EN
    2:   IO3 RX0     GND
    3:   IO0         3V3
    4:   GND         EN
    5:   IO39        IO32 CFG
    6:   IO36        IO33 485_EN
    7:   IO15        IO5  RXD
    8:   IO14        IO17 TXD
    9:   IO12        GND
    10:  IO35        3V3
    11:  IO4         GND
    12:  IO2         5V
    13:  GND         LINK
```
Gpio 5/17 do have small leds, but this should not preclude their usage.

**Comment by shikousa** [comment]
Ok. Then I will try the input-only PINs 35,36,39 for RX. I will connect this PIN to an ADUM1201 for levelshifting 5V <-> 3V. As the input-only PINs have no internal pull-up, do you know if I need an external pull-up/down to connect it to an ADUM1201 or is it also possible without?

**Reply by sfromis** [reply]
No, I have no knowledge of that thing. In general, I'd "start small" and get it to work on "safe" pins, one thing at a time. Only after that, go on to other pins, like in case of wanting pins for other stuff.
